### ISSUE TITLE
RFC: everything in rust docs

### ISSUE THREAD

**Issue by sam0x17** [paritytech/polkadot-sdk#991]
This came up in our FRAME weekly meeting, some internal docs discussions, and some conversations with community members on discord and elsewhere, and across the board there has been wide support for it so re-posting here for a more public discussion.

## Everything in Rust Docs

Tutorials, documentation explaining what various blockchain and FRAME concepts are..._everything_... can and should live in rust-docs in our (soon-to-be-a-thing) monorepo because this is the lowest friction way for people to learn about these things (inline, at the exact moment when their confusion occurs while coding).

The way this could work is with a `substrate-concepts` / `frame-concepts` or something like that sub-crate (probably dev dependency of `frame` crate or root `substrate` crate) that is _docs-only_ and has a bunch of empty modules (or ZSTs) that correspond with various concepts, tutorials, etc. The rust docs for these modules would be the content of tutorials or explanations of concepts, etc., and can use a `docify`-like approach to embed markdown files for the longer ones especially tutorials that are easier to maintain as a .md file. Images we can easily embed, videos would become links which I think is fine.

Then for example anywhere you mention something related to "staking", you could directly doc-link to [`Staking`](`frame_concepts::Staking`), and since these are dev-only dependencies, we can link to them from anywhere in substrate without causing dependency-related strain.

The exact setup I'm happy to iterate on with a working example that maximizes what we can docify-embed, but this is the general idea.

## Advantages

* If all of these things are version-controlled under the same repo, we solve the extremely difficult problem of out-dated/semver-tagged tutorial/concept/etc links.
* If everything is in the same repo, we can use tools like `docify` to directly embed code examples from elsewhere in substrate _within the tutorials that reference them_ such that they will automatically be updated when they are changed in the code.
* Once we start publishing the monorepo to crates.io, all of our docs will get massive SEO and google visibility from docs.rs, lib.rs and all the docs.rs clone sites, much better than we could ever do with our own docs domains because of how pagerank works. This also means we don't need devops/etc to maintain these sites, we can just lean on crates.io to publish everything.
* Search within our rust docs becomes like a superpower that will tell you anything you want to know about our ecosystem / blockchain technology / etc.
* Common blockchain and substrate/FRAME concepts become significantly more discoverable/learnable when browsing our rust docs.
* Developers have direct line-of-sight into all of our docs without going to a different repo and can update them as part of any PR that makes changes to substrate.
* Doc changes are published in lock-step with code changes that those docs relate to.

Would love some thoughts on this and further discussion. General reception so far has been highly positive so would love to hear some dissenting views.

cc @kianenigma @NukeManDan @liamaharon @gupnik @franciscoaguirre @KiChjang @ggwpez @juangirini @aaronbassett

**Comment by gilescope** [comment]
When we go to our new doc portal, (wherever the new version of https://paritytech.github.io/cumulus/ goes to) I'd like it if private documentation is turned on - that's not available in docs.rs and is helpful. E.g. documentation around nft pallet features is all non-public at the moment.

**Comment by KiChjang** [comment]
The one thing that I don't quite completely like about this approach is the purported "docs situated right next to code" aspect -- this is not quite true. Sure, we can have references and links to actual types, but the reality is that the docs are hosted in a completely different `doc` module altogether, so it's at least a layer removed from the actual code.

That does change a couple of things about the advantages listed out here:
* We could still get an out-of-date documentation, simply because we forgot to look in the `doc` module and update the text there;
* Doc changes may not necessarily be published in lock-step with code changes;
* Devs don't really have a direct line-of-sight, unless we purposefully introduce links back to the `doc` module from the concrete types/structs/functions

**Comment by liamaharon** [comment]
I'm writing new Migration and Try State tutorial/docs (which could find their way into the frame master tutorial when that is ready), and was initially planning to experiment writing this tutorial/guide format in Rustdocs. 

However, after working on it I feel like mdbook is better suited for stepped-tutorials.
- Prettier, easier read
- Better navigation UX
- The ToC on the left hand side of mdbook is easier for the user to navigate through sections, and easier for the developer to write & organise sections than trying to create a new crate for every 'section' in rustdocs.

Everything is still just markdown, so it won't be difficult to migrate to rustdocs in the future if we decide.

I'm still in support of experimenting with all standalone pieces of docs that don't belong in a stepped tutorial being in rustdocs.

**Comment by sam0x17** [comment]
> * Devs don't really have a direct line-of-sight, unless we purposefully introduce links back to the `doc` module from the concrete types/structs/functions

Yeah, doubling-down on doing this everywhere would be key

**Comment by sam0x17** [comment]
> However, after working on it I feel like mdbook is better suited for stepped-tutorials.

Yeah, this makes me wonder if there might be some way to export this > mdbook or mdbook > this in some automated way where we can get the best of both worlds

**Comment by chevdor** [comment]
## All rustdoc 

While I like the idea, I am not convinced using exclusively rustdoc would work for **all authors** (especially non dev ones). There will always be markdown written and not always embedded in some rustdoc.

As far as code is concerned and direct documentation of the code, afaic there is no doubt this should be rustdoc in the code or linked markdown.

The topic rustdoc vs markdown book will depend of the reader. 
Someone coding will prefer rustdoc where **all** the documentation of a given piece of code is available, whereas for a  general reader, a "book" will be better to extract relevant content (and skip less relevant one).

## mdbook / static site generator

As for the more generic documentation, I would leave it to the doc team to pick the the solution they can be maintained best.

I see a big benefit in bring all the doc we have together in one browsable/searchable site. That will also contribute keeping a good quality of content.

> Everything is still just markdown, so it won't be difficult to migrate to rustdocs in the future if we decide.

I agree as long as links are valid and it may be problematic in "flying" markdown that may not always be in sync with some pieces of code it refers to.

## no dup

It is obvious and critical to have no duplication in the doc source, we should not have content in both code (.rs) and markdown and no markdown duplicated for the sake of generating rustdoc and some markdown book.

That being said, I don't think it would be an issue to have part of the doc used in both some rustdoc and some markdownbook/statuic doc site.

## URL validity

There are related issues:
- #993 
- #134

The later is the first to be taken care of, then we can enable a fix for #993 to ensure URLs remain valid. That will work for external urls.

## Documentation decay

One reality is that some part of the doc become stale over time. This can be due to maintainers of a piece of code to move of to other topics, not being around anymore, or the lack of updates due to oversights. 

`rustdoc` has the edge here since it will help keep the doc in sync with the code whereas some "flying" markdown could get out of sync.

Our `CODEOWNERS` config was no longer valid in the monorepo, I prepared a fix: #1415

I was also tinkering with the idea of adding a "freshness/anti-decay check" to the markdown at least.

So for instance, a `.md` file not updated for more than N days (we probably would start with some value in months...) would become invalid and raise a **warning** in CI.

If is still valid, a simple noop change could reset the counter and mark the validation that the doc was checked and is "fresh" enough.

It is never fun to take over doc written by others but this will/is become necessary to ensure the overall quality.

**Comment by chevdor** [comment]
see also https://paritytech.github.io/polkadot-sdk/

**Comment by sam0x17** [comment]
> The topic rustdoc vs markdown book will depend of the reader.

The nice thing about using `docify` here is you can have both -- we can have a folder of markdown files that becomes nested modules of rustdoc _and_ gets published somewhere as an md book

**Comment by kianenigma** [comment]
This is a great idea, I will adopt it as such: 

In the new doc-vision manifesto, there are 3 classes of documents that we produce: 

1. A master tutorial
2. Native rust-docs for low level items, functions and Rust items.
3. Reference docs for whatever remains in between that it too broad to be documented in rust-doc, and too specific to be in the master tutorial.

(See [here](https://forum.parity.io/t/new-polkadot-substrate-documentation-plan/1810#action-items-4) for a more detailed description).

The intention of the docs effort should be to do as much of the above as possible in rust-docs. As noted, these rust-docs could use `docify`, be processed into `.md` files, which will itself become a `md-book` of some sort. 

But the main point is, the source of truth is always `.rs` files.
